tensorflow/python/grappler/auto_mixed_precision_test.py

```diff
@@ -0,0 +1,555 @@
+  o = _weight([6])
+  y, _, _ = _fused_batchnorm(x, s, o)
+  y = array_ops.identity(y)
+  return y
+
+
+def _matmul_act(x):
+  """Matmul followed by activation."""
+  i = array_ops.reshape(x, [8, 8])
+  f = _weight([8, 8])
+  x = math_ops.matmul(i, f)
+  y = nn.relu(x)
+  return y
+
+
+def _conv_pool(x):
+  """Conv followed by pooling."""
+  x_image = array_ops.reshape(x, [-1, 8, 8, 1])
+  w_conv1 = _weight([3, 3, 1, 6])
+  b_conv1 = _bias([6])
+  h_conv1 = nn.relu(nn.bias_add(_conv2d(x_image, w_conv1), b_conv1))
+  h_pool1 = _max_pool_2x2(h_conv1)
+  w_conv2 = _weight([3, 3, 6, 4])
+  b_conv2 = _bias([4])
+  h_conv2 = nn.relu(nn.bias_add(_conv2d(h_pool1, w_conv2), b_conv2))
+  h_pool2 = _max_pool_2x2(h_conv2)
+  return h_pool2
+
+
+def _simple_loop(x, functor):
+  """Simple loop whose body is provided by the functor."""
+  init = (constant_op.constant(0), x)
+  c = lambda i, j: i < 4
+  b = lambda i, j: (i+1, functor(j))
+  ij = control_flow_ops.while_loop(c, b, init)
+  return ij
+
+
+
+def _loop_vars_intertwined(x0, y0, functor_x, functor_y):
+  """Loop whose loop variables are intertwined."""
+  c = lambda i, j, x, y: j < 4
+  b = lambda i, j, x, y: (j+1, i+1, functor_y(y), functor_x(x))
+  init = (constant_op.constant(0), constant_op.constant(0), x0, y0)
+  ijzw = control_flow_ops.while_loop(c, b, init)
+  return ijzw
+
+
+def _lstm_cell(prev_c, prev_h, x):
+  """ LSTMCell
+  i: input gate
+  f: forget gate
+  o: output gate
+  c: cell state
+  x: input
+  h: embedding
+  """
+  bias = _bias([4])
+  w = _weight([8, 16])
+  ifoc = math_ops.matmul(array_ops.concat([x, prev_h], axis=1), w)
+  i, f, o, c = array_ops.split(ifoc, 4, axis=1)
+  i = math_ops.sigmoid(nn.bias_add(i, bias))
+  f = math_ops.sigmoid(nn.bias_add(f, bias))
+  o = math_ops.sigmoid(nn.bias_add(o, bias))
+  c = math_ops.tanh(nn.bias_add(c, bias))
+  next_c = f * prev_c + i * c
+  next_h = o * math_ops.tanh(next_c)
+  return next_c, next_h
+
+
+def _recurrent_lstm(c, h):
+  """ Dynamic single-layer LSTM with TensorArray """
+  def cond(i, c, h, ta_x):
+    del c
+    del h
+    del ta_x
+    return i < 4
+
+  def body(i, c, h, ta_x):
+    x = ta_x.read(i)
+    next_c, next_h = _lstm_cell(c, h, x)
+    return (i+1, next_c, next_h, ta_x)
+
+  ta_x = tensor_array_ops.TensorArray(
+      dtype=dtypes.float32,
+      size=4)
+  for i in range(0, 4):
+    ta_x = ta_x.write(
+        i, constant_op.constant(0.1, shape=[8, 4],
+                                dtype=dtypes.float32))
+  init = (constant_op.constant(0), c, h, ta_x)
+  r = control_flow_ops.while_loop(cond, body, init)
+  return r
+
+
+def _make_node_with_color(color, input_tensor, name=None):
+  """ Returns a node representative of the specified list type """
+  color = color.lower()
+  if color == 'w': # White node
+    weights = _weight(input_tensor.get_shape().as_list())
+    return math_ops.matmul(input_tensor, weights, name=name)
+  if color == 'g': # Gray node
+    return math_ops.sqrt(input_tensor, name=name)
```

Comment:
Having sqrt and log is fine, but will cause NaNs on negative numbers, and in tests, you assert variables are all close. NaN is close to NaN but that's not a useful check, so change these to values that do not cause NaNs, or add a TODO.

Reply:
You're right, these tests were producing NaNs and silently failing.
I've changed them to NaN-proof ops.